feat: replace google analytics with custom referral tracking and storage

Ekep-Obasi opened the pull request:
Resolves #74

<!--
- You must link the issue number e.g. "Resolves #1234"
- Please do not replace the keyword "Resolves" https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

Ekep-Obasi committed:
feat: replace google analytics with custom referral tracking and storage

0x4007 reviewed (commented):
Code looks good. @gentlementlegen you're pretty good with supabase what do you think about this?

gentlementlegen commented:
Would be nice to:
- have CI tests pass
- have an Action that generates Supabase types and commits them so contributors that do not have access to the database can have up to date types
- add related tests
- use Ubiquity's logger instead of console logs

0x4007 commented:
> Would be nice to:
> 
> * have CI tests pass

None of the code changes stand out to me as causing Cypress to fail. Perhaps the test is unreliable?

gentlementlegen commented:
Error is here
```
  1) DevPool
       Items can be sorted - bottom row - portrait/mobile:
     CypressError: Timed out retrying after 4000ms: `cy.scrollTo()` failed because this element is not scrollable:

`<div id="filters-bottom" class="filters-container">...</div>`

Make sure you're targeting the correct element or use `{ensureScrollable: false}` to disable the scrollable check.
```

Maybe this error was here before.

Ekep-Obasi committed:
qadded tests

Ekep-Obasi committed:
added tests

gentlementlegen commented on cypress/e2e/devpool.cy.ts:
```diff
@@ -250,6 +250,49 @@ describe("DevPool", () => {
+      });
+    }).as("getUser");
+
+    cy.intercept("POST", "https://lelwqeowubwpkhjlzsiu.supabase.co/rest/v1/referrals", (req) => {
```
Ideally this should not be hardcoded.

gentlementlegen commented on cypress/e2e/devpool.cy.ts:
```diff
@@ -199,7 +199,7 @@ describe("DevPool", () => {
     cy.get('[for="priority-bottom"]').click();
     cy.get('div[id="issues-container"]').children().should("have.length", 2);
-    cy.get("#filters-bottom").scrollTo("right");
+    cy.get("#filters-bottom").scrollTo("right", {ensureScrollable: false});
```
Doesn't seem to respect prettier rules.

Ekep-Obasi replied:
Thanks @gentlementlegen, let me do the changes, Could you approve the workflows

gentlementlegen replied:
I think the workflows are working for each commit?

0x4007 replied:
> I think the workflows are working for each commit?

I ran it manually

Ekep-Obasi committed:
ci: workflow to generate and commit supabase types

Ekep-Obasi committed:
chore: removed hardcoded values in test

Ekep-Obasi committed:
chore: removed hardcoded values in test

Ekep-Obasi committed:
chore: changed console logs to ubiquity logs

Ekep-Obasi committed:
chore: prettier format

gentlementlegen commented on .github/workflows/supabase-types.yml:
```diff
@@ -0,0 +1,47 @@
+    permissions:
+      contents: write
+    env:
+      PROJECT_REF: "lelwqeowubwpkhjlzsiu"
```
Should come from the environment

gentlementlegen commented on .github/workflows/supabase-types.yml:
```diff
@@ -0,0 +1,47 @@
+name: Update database types
+
+on:
+  push:
```
I think this should happen only on the main branch, so it's the latest types and it doesn't bombard the repo with commits.

gentlementlegen commented on .github/workflows/supabase-types.yml:
```diff
@@ -0,0 +1,47 @@
+        with:
+          persist-credentials: false
+          fetch-depth: 0
+      - uses: actions/setup-node@v2.1.5
```
Why this version specifically?

Ekep-Obasi committed:
fix: update supabase type script

gentlementlegen commented on .github/workflows/supabase-types.yml:
```diff
@@ -0,0 +1,47 @@
+    env:
+      PROJECT_REF: "lelwqeowubwpkhjlzsiu"
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v4
```

gentlementlegen commented on .github/workflows/supabase-types.yml:
```diff
@@ -0,0 +1,47 @@
+          git commit -m 'chore: update Supabase types'
+          git push
+
+      - name: Push changes
```
You use this step but the latest command of the previous step seems to be `git push`?

gentlementlegen commented on supabase/tables/base.ts:
```diff
@@ -0,0 +1,12 @@
+
+  constructor(supabase: SupabaseClient) {
+    this.supabase = supabase;
+    this.logger = {} as Logs;
```
Why not `new Logger()`?

Ekep-Obasi replied:
I avoided `new Logger()` to keep the constructor simple and avoid unnecessary configuration which are not optional. Instead, I initialized logger as an empty object for simplicity.

gentlementlegen replied:
So are the inheritors responsible to initialize the logger? What happens if a class does not initialize it?

Ekep-Obasi replied:
You're right. Inheritors need to initialize the logger, or it will stay as an empty object and could cause issues.

gentlementlegen commented on .github/workflows/supabase-types.yml:
```diff
@@ -0,0 +1,47 @@
+        run: yarn install
+
+      - name: Generate Supabase types
+        run: npm run update-types
```
Probably should use `yarn`

0x4007 commented on supabase/tables/base.ts:
```diff
@@ -0,0 +1,12 @@
+import { SupabaseClient } from "@supabase/supabase-js";
+import { Logs } from "ubiquibot-logger";
+
+export class SuperBase {
```
```suggestion
export class SupabaseClient {
```

Maybe this is more accurate.

0x4007 commented on supabase/config.toml:
```diff
@@ -0,0 +1,178 @@
+enable_signup = true
+# If enabled, users need to confirm their phone number before signing in.
+enable_confirmations = false
+# Template for sending OTP to users
+template = "Your code is {{ .Code }} ."
+# Controls the minimum amount of time that must pass before sending another sms otp.
+max_frequency = "5s"
+
+# Use pre-defined map of phone number to OTP for testing.
+# [auth.sms.test_otp]
+# 4152127777 = "123456"
+
+# This hook runs before a token is issued and allows you to add additional claims based on the authentication method used.
+# [auth.hook.custom_access_token]
+# enabled = true
+# uri = "pg-functions://<database>/<schema>/<hook_name>"
+
+# Configure one of the supported SMS providers: `twilio`, `twilio_verify`, `messagebird`, `textlocal`, `vonage`.
+[auth.sms.twilio]
+enabled = false
+account_sid = ""
+message_service_sid = ""
+# DO NOT commit your Twilio auth token to git. Use environment variable substitution instead:
+auth_token = "env(SUPABASE_AUTH_SMS_TWILIO_AUTH_TOKEN)"
+
+# Use an external OAuth provider. The full list of providers are: `apple`, `azure`, `bitbucket`,
+# `discord`, `facebook`, `github`, `gitlab`, `google`, `keycloak`, `linkedin_oidc`, `notion`, `twitch`,
+# `twitter`, `slack`, `spotify`, `workos`, `zoom`.
+[auth.external.apple]
+enabled = false
+client_id = ""
+# DO NOT commit your OAuth provider secret to git. Use environment variable substitution instead:
+secret = "env(SUPABASE_AUTH_EXTERNAL_APPLE_SECRET)"
+# Overrides the default auth redirectUrl.
+redirect_uri = ""
+# Overrides the default auth provider URL. Used to support self-hosted gitlab, single-tenant Azure,
+# or any other third-party OIDC providers.
+url = ""
+# If enabled, the nonce check will be skipped. Required for local sign in with Google auth.
+skip_nonce_check = false
```
```suggestion
```
Perhaps there are some unnecessary configs.

Ekep-Obasi replied:
I’m not very familiar with this, but I could clean up some of the comments

Ekep-Obasi committed:
fix: typo, rename SuperBase class to Supabase

Ekep-Obasi committed:
ci: update workflow to use GitHub secrets for environment variables

Ekep-Obasi committed:
chore: remove comments for a cleaner config file

gentlementlegen commented on package.json:
```diff
@@ -17,7 +17,8 @@
     "cypress:run": "cypress run",
-    "cypress:headed": "cypress run --headed"
+    "cypress:headed": "cypress run --headed",
+    "update-types": "npx supabase gen types --lang=typescript --project-id $(grep -oP '(?<=^PROJECT_ID=).+' .env) > supabase/types.ts"
```
`grep` is not cross environment friendly, and does this handle local / remote environments?

Ekep-Obasi replied:
Yeah that true, It'll be preferable to write a script

gentlementlegen replied:
Using a script is fine but maybe overkill, you can just use `cross-env` and `dotenv` to load the variable within the command line (and it works on any environment). Also, would be nice to have a `local` and `remote` command to generate the types either locally with a local db or remotely targeting a deployed database.

gentlementlegen commented on package.json:
```diff
@@ -52,6 +53,7 @@
     "npm-run-all": "^4.1.5",
     "prettier": "^3.1.0",
     "tsx": "^4.6.0",
+    "ubiquibot-logger": "^0.3.5",
```
https://www.npmjs.com/package/@ubiquity-dao/ubiquibot-logger should be the proper package.

Ekep-Obasi committed:
fix: switched to right logger dependency

Ekep-Obasi commented:
@gentlementlegen, How are the changes looking now? Are there other areas where I could make improvements?

gentlementlegen commented on package.json:
```diff
@@ -17,7 +17,8 @@
     "cypress:run": "cypress run",
-    "cypress:headed": "cypress run --headed"
+    "cypress:headed": "cypress run --headed",
+    "update-types": "dotenv -e .env -- sh -c 'npx supabase gen types --lang=typescript --project-id $PROJECT_ID > supabase/types.ts'"
```
`sh` is also not cross platform. Please have a look at https://www.npmjs.com/package/cross-env

Ekep-Obasi replied:
From what I see here,`cross-env` is good for setting environment variables in a cross-platform way but does not handle `.env` files.

gentlementlegen replied:
Correct which is why you should combine `cross-env-shell` to run the script and `dotenv` to load the environment file.

Ekep-Obasi replied:
I went through `cross-env-shell` package, I could not really find anything helpful https://www.npmjs.com/package/cross-env-shell.

But I went for another similar package `dotenv-cli`, Using it with `bash -c` in GitHub Actions, which runs on Ubuntu should work

gentlementlegen commented on supabase/config.toml:
Does this imply we need to spin up a new Supabase database?

Ekep-Obasi replied:
yeah, I think so

Ekep-Obasi committed:
fix: failing tests and clean ups

Ekep-Obasi commented:
@gentlementlegen Could you please approve the workflows?

Ekep-Obasi commented:
Hey @gentlementlegen, @0x4007, @Keyrxng,

I wanted to follow up on my pull request from last week. When you get a chance, I would appreciate it if someone could take a look and approve it if everything looks good.

Keyrxng commented on supabase/seed.sql:
shouldn't this contain a script to setup the DB? Or there should be a migration script instead but I don't see the value in an empty file here

Keyrxng commented on supabase/types.ts:
We usually avoid committing these I'm sure. Is this strictly needed? Could a simple type be used instead to represent the table?

Keyrxng replied:
@gentlementlegen rfc

edit:


I just saw [your comment](https://github.com/ubiquity/work.ubq.fi/pull/75#issuecomment-2246858755) and to follow on

> Does this imply we need to spin up a new Supabase database?
> yeah, I think so

so are we spinning up a new DB just for referrals or just adding a table to the main DB? Either way I'd have thought if we are rebuilding the referral system from scratch then there would be no breaking changes to come for a long time and simple types would be better

Just my opinion, I think the Supabase types are unattractive and can be overwhelming but them being dynamic is ofc a bonus

gentlementlegen replied:
I think it is good and bad to commit these, but so far we always did.
Pros:
- any developer can clone and use
- maybe developers do not have access to the DB and it's nice to have the types without spinning up a local db

Cons:
- pollutes the repo
- easily can merge conflict
- knip will complain on missing types

I think the best scenario is to have an Action generating the types on any push to the main branch as it ensures that the deployed db and the code are up to date with each other. Either way there has to be some script generating these types, either during build or during a push to the main branch.

Keyrxng commented on supabase/config.toml:
Can this config be stripped down to only what we need to use?

Keyrxng replied:
This is still unresolved. 

Ask yourself what features of this have you been using while developing? Out of curiosity, are you using a local or remote Supabase instance and did you setup a Github app for logging in and testing with or did you use the ubiquity DB for auth? 

How is the app connected to Supabase and what auth strategy does the UI implement? What is it we are using the storage api and buckets for, are they required? Do we login with twillio or sms or email? Do we need S3 env keys? 

Likely most, if not all of these can be removed. I'm no expert with Supabase but what do we do with this file within the context of this repo/feature? Obviously it's for configuring a DB but where/how is it used?

Keyrxng commented on src/home/devrel-tracker.ts:
```diff
@@ -1,22 +1,30 @@
+    const devRelCode = urlParams.get("ref");
     if (devRelCode) {
-      localStorage.setItem("devRel", devRelCode);
+      localStorage.setItem("ref", devRelCode);
```
> One thing that I Just noticed is that in the URL https://work.ubq.fi/?devRel=XYZ the query parameter devRel is case-sensitive. The R in devRel must be capitalized. It would have been nice to have it case-insensitive but this is what is working now. 

A comment from the original implementation PR @gentlementlegen is this something we should consider? I'm unsure what the referral url schema looks like beyond `ref=`

0x4007 replied:
I think `ref` is sufficient for our needs here.

Keyrxng replied:
sorry @0x4007 I meant what comes after `ref=`, the code or string or whatever the schema is, is case sensitivity a concern there? I presume no with the above but I haven't seen any campaign link generating first-hand so I wanted to check, ty.

Keyrxng reviewed (changes requested):
The localStorage and posting to Supabase logic works

I setup a new database and had to disable RLS to allow me to push. RLS should probably be defined for this table right?

I'd suggest for the authenticated user and that they can only edit their own entry, probably reads should be disabled as well

This can be done as part of the migration/setup script

Keyrxng commented:
@0x4007 with what I assume is more freedom with what can be stored in the DB vs GTAG events should this be given some thought as to how we store things

> Set up a storage solution
> Record referrals.

e.g do we need `dev_github = "username|userId"` or does it improve things if they are separate rows? Could the primaryKey be the `userId` to easily enforce uniqueness/once-per and authenticated-user RLS edits?

Perhaps it's best that records cannot be edited or seen at all once posted to the DB?

I'll admit I'm not fantastic with architecting backend storage solutions at the best of times but definitely not when practically off the top of my head

0x4007 commented:
All data should be in different columns as much as possible to make querying it as accurate as possible.

Ekep-Obasi committed:
fix: merge conflicts

Ekep-Obasi committed:
refactor: separate github fields and simplify config

    - split github username and user id in devrel tracking
    - update tests for new schema
    - simplify supabase config by removing unnecessary settings
    - remove empty seed.sql

gentlementlegen commented on package.json:
```diff
@@ -17,7 +17,7 @@
     "prepare": "husky install",
     "cypress:open": "cypress open",
     "cypress:run": "cypress run",
-    "cypress:headed": "cypress run --headed"
```
Why was cypress removed?

Ekep-Obasi replied:
my bad, I must have removed this by mistake

gentlementlegen commented on package.json:
```diff
@@ -17,7 +17,7 @@
     "cypress:open": "cypress open",
     "cypress:run": "cypress run",
-    "cypress:headed": "cypress run --headed"
+    "update-types": "dotenv -e .env -- bash -c 'npx supabase gen types --lang=typescript --project-id $PROJECT_ID --schema public > supabase/types.ts'"
```
Thank you for this. But please make it cross platform (this would not work outside of bash environments). Please refer to `cross-env` or equivalent.

gentlementlegen commented on package.json:
```diff
@@ -41,7 +43,7 @@
     "@typescript-eslint/eslint-plugin": "^6.13.1",
     "@typescript-eslint/parser": "^6.13.1",
-    "cypress": "13.11.0",
+    "cypress": "^13.11.0",
```
```suggestion
    "cypress": "13.11.0",
```

gentlementlegen commented on supabase/tables/base.ts:
```diff
@@ -0,0 +1,12 @@
+import { SupabaseClient } from "@supabase/supabase-js";
+import { Logs, LOG_LEVEL, LogLevel } from "@ubiquity-dao/ubiquibot-logger";
+
+export class SupaBase {
```
```suggestion
export class Supabase {
```
Except if this names collides with something.

Ekep-Obasi committed:
fix: types update scripts

Keyrxng commented on package.json:
```diff
@@ -17,7 +17,8 @@
     "cypress:run": "cypress run",
-    "cypress:headed": "cypress run --headed"
+    "cypress:headed": "cypress run --headed",
+    "update-types": "cross-env dotenv -e .env -- bash -c 'npx supabase gen types --lang=typescript --project-id $PROJECT_ID --schema public > supabase/types.ts'"
```
Without a migration file or a way to setup a DB programmatically, if a new contributor wanted to use their own DB they'd need to root around your code/types and manually create it first.

Failing that they'd erase your types and the logic would fail as the DB would not be ready to accept the data. They'd also have to create RLS policies too, which would typically be a part of the migration file.

What is your solution to this problem?

Keyrxng commented on package.json:
```diff
@@ -30,7 +31,10 @@
     "@octokit/request-error": "^6.1.0",
     "@octokit/rest": "^20.0.2",
     "@supabase/supabase-js": "^2.39.0",
+    "@ubiquity-dao/ubiquibot-logger": "^1.3.0",
```
can you bump this to `1.3.1` please?

Keyrxng commented on supabase/types.ts:
@gentlementlegen double checking here that we are definitely committing this file?

If we use the main DB it exposes everything and we usually avoid that. If we spin up a DB dedicated to referrals it means more infra to manage and less interoperable with the main DB.

gentlementlegen replied:
We are committing this to other projects, and we need it for Knip not to complain. You can always introspect Supabase afaik, as long as you provide an ANON key (try it out here https://supabase-schema.vercel.app/). If so then it should be fine.

Keyrxng reviewed (changes requested):
Thinking in terms of future contributions, they will need to use their own DB instance as the production DB should have RLS that prevents dangerous actions on it which would make testing anything to do with the feature difficult if using the main DB details.

This means manually creating a new DB/new tables every time unless we have a migration file included as part of this PR.

I think the readme should be updated to include steps on how a new contributor would setup the DB and what's needed as part of this new feature.

Ekep-Obasi commented:
/start
